### PR TITLE
planner: support null-aware semi join push down to tiflash | tidb-test=pr/2102

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: gengliqi <gengliqiii@gmail.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/40745

Problem Summary:
Support `Null-aware anti left outer semi join` and `Null-aware anti semi join` push down to tiflash.

### What is changed and how it works?
```sql
mysql> explain select a,b,(a,b) not in (select a,b from s where s.c >= t.c) from t;
+----------------------------------------+---------+--------------+---------------+--------------------------------------------------------------------------------------------------------------------------------+
| id                                     | estRows | task         | access object | operator info                                                                                                                  |
+----------------------------------------+---------+--------------+---------------+--------------------------------------------------------------------------------------------------------------------------------+
| TableReader_28                         | 3.00    | root         |               | data:ExchangeSender_27                                                                                                         |
| └─ExchangeSender_27                    | 3.00    | mpp[tiflash] |               | ExchangeType: PassThrough                                                                                                      |
|   └─Projection_9                       | 3.00    | mpp[tiflash] |               | test.t.a, test.t.b, Column#13                                                                                                  |
|     └─HashJoin_26                      | 3.00    | mpp[tiflash] |               | Null-aware anti left outer semi join, equal:[eq(test.t.b, test.s.b) eq(test.t.a, test.s.a)], other cond:ge(test.s.c, test.t.c) |
|       ├─ExchangeReceiver_15(Build)     | 2.00    | mpp[tiflash] |               |                                                                                                                                |
|       │ └─ExchangeSender_14            | 2.00    | mpp[tiflash] |               | ExchangeType: Broadcast                                                                                                        |
|       │   └─TableFullScan_13           | 2.00    | mpp[tiflash] | table:s       | keep order:false, stats:pseudo                                                                                                 |
|       └─TableFullScan_12(Probe)        | 3.00    | mpp[tiflash] | table:t       | keep order:false, stats:pseudo                                                                                                 |
+----------------------------------------+---------+--------------+---------------+--------------------------------------------------------------------------------------------------------------------------------+
8 rows in set (0.01 sec)
```

```sql
mysql> explain select a,b from t where (a,b) not in (select a,b from s where s.c >= t.c);
+----------------------------------------+----------+--------------+---------------+---------------------------------------------------------------------------------------------------------------------+
| id                                     | estRows  | task         | access object | operator info                                                                                                       |
+----------------------------------------+----------+--------------+---------------+---------------------------------------------------------------------------------------------------------------------+
| TableReader_28                         | 8000.00  | root         |               | data:ExchangeSender_27                                                                                              |
| └─ExchangeSender_27                    | 8000.00  | mpp[tiflash] |               | ExchangeType: PassThrough                                                                                           |
|   └─Projection_9                       | 8000.00  | mpp[tiflash] |               | test.t.a, test.t.b                                                                                                  |
|     └─HashJoin_26                      | 8000.00  | mpp[tiflash] |               | Null-aware anti semi join, equal:[eq(test.t.b, test.s.b) eq(test.t.a, test.s.a)], other cond:ge(test.s.c, test.t.c) |
|       ├─ExchangeReceiver_15(Build)     | 10000.00 | mpp[tiflash] |               |                                                                                                                     |
|       │ └─ExchangeSender_14            | 10000.00 | mpp[tiflash] |               | ExchangeType: Broadcast                                                                                             |
|       │   └─TableFullScan_13           | 10000.00 | mpp[tiflash] | table:s       | keep order:false, stats:pseudo                                                                                      |
|       └─TableFullScan_12(Probe)        | 10000.00 | mpp[tiflash] | table:t       | keep order:false, stats:pseudo                                                                                      |
+----------------------------------------+----------+--------------+---------------+---------------------------------------------------------------------------------------------------------------------+
8 rows in set (0.01 sec)
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -260,5 +260,6 @@ replace (
 	// fix potential security issue(CVE-2020-26160) introduced by indirect dependency.
 	github.com/dgrijalva/jwt-go => github.com/form3tech-oss/jwt-go v3.2.6-0.20210809144907-32ab6a8243d7+incompatible
 	github.com/pingcap/tidb/parser => ./parser
+	github.com/pingcap/tipb => github.com/gengliqi/tipb v0.0.0-20230119154529-60cd322dc122
```
Comment (hawkingrei) [inline go.mod:263]: Please run ```make bazel_prepare```

File: planner/core/casetest/testdata/expression_rewriter_suite_out.json [hunk 2]
```diff
@@ -21,7 +21,7 @@
       {
         "SQL": "select * from t1 where (t1.a, t1.b) not in (select a, b from t2)",
         "Plan": [
-          "HashJoin 3.20 root  CARTESIAN anti semi join, other cond:eq(test.t1.a, test.t2.a), eq(test.t1.b, test.t2.b)",
+          "HashJoin 3.20 root  Null-aware anti semi join, equal:[eq(test.t1.b, test.t2.b) eq(test.t1.a, test.t2.a)]",
```
Comment (windtalker) [inline planner/core/casetest/testdata/expression_rewriter_suite_out.json:24]: Why the key order is changed

File: planner/core/exhaust_physical_plans.go [hunk 3]
```diff
@@ -2006,7 +2006,7 @@ func (p *LogicalJoin) shouldUseMPPBCJ() bool {
 
 // canPushToCop checks if it can be pushed to some stores.
 func (p *LogicalJoin) canPushToCop(storeTp kv.StoreType) bool {
-	return len(p.NAEQConditions) == 0 && p.baseLogicalPlan.canPushToCop(storeTp)
+	return p.baseLogicalPlan.canPushToCop(storeTp)
```
Comment (windtalker) [inline planner/core/exhaust_physical_plans.go:2009]: maybe we can just delete this function?

File: planner/core/plan_to_pb.go [hunk 4]
```diff
@@ -416,15 +416,29 @@ func (p *PhysicalIndexScan) ToPB(_ sessionctx.Context, _ kv.StoreType) (*tipb.Ex
...
 	client := ctx.GetClient()
-	// todo: mpp na-key toPB.
-	leftJoinKeys := make([]expression.Expression, 0, len(p.LeftJoinKeys))
-	rightJoinKeys := make([]expression.Expression, 0, len(p.RightJoinKeys))
-	for _, leftKey := range p.LeftJoinKeys {
-		leftJoinKeys = append(leftJoinKeys, leftKey)
-	}
-	for _, rightKey := range p.RightJoinKeys {
-		rightJoinKeys = append(rightJoinKeys, rightKey)
+	var leftJoinKeys, rightJoinKeys []expression.Expression
```
Comment (windtalker) [inline planner/core/plan_to_pb.go:419]: throw error if both `LeftNAJoinKeys` and `LeftJoinKeys` are not empty?

File: planner/core/casetest/testdata/expression_rewriter_suite_out.json (hunk 2 above)
Comment (gengliqi) [reply planner/core/casetest/testdata/expression_rewriter_suite_out.json:24]: Seems the order has all changed. @AilinKid any idea?

File: planner/core/exhaust_physical_plans.go (hunk 3 above)
Comment (gengliqi) [reply planner/core/exhaust_physical_plans.go]: Yes, removed.

File: planner/core/plan_to_pb.go (hunk 4 above)
Comment (gengliqi) [reply planner/core/plan_to_pb.go:419]: Done.

File: planner/core/casetest/testdata/expression_rewriter_suite_out.json (hunk 2 above)
Comment (gengliqi) [reply planner/core/casetest/testdata/expression_rewriter_suite_out.json]: I find the root cause and fix it.

Comment (AilinKid) [reply planner/core/casetest/testdata/expression_rewriter_suite_out.json]: I disorder the keys, just a mistake, but won't affect the result

File: planner/core/casetest/testdata/enforce_mpp_suite_in.json [hunk 5]
```diff
@@ -158,5 +158,19 @@
...
       "explain select count(distinct a), count(distinct b), avg(c+d) from t"
     ]
+  },
+  {
+    "name": "TestMPPNullAwareSemiJoinPushDown",
+    "cases": [
+      "set @@tidb_allow_mpp=1;set @@tidb_enforce_mpp=1;set @@tidb_enable_null_aware_anti_join=on;",
+      "EXPLAIN select * from t where t.a not in (select s.a from s); -- 1. anti semi join, one join key",
+      "EXPLAIN select * from t where t.a not in (select s.a from s where t.c > s.c); -- 2. anti semi join, one join key + other condition",
+      "EXPLAIN select * from t where (t.a, t.b) not in (select s.a, s.b from s); -- 3. anti semi join, two one join keyd",
```
Comment (AilinKid) [inline planner/core/casetest/testdata/enforce_mpp_suite_in.json:168]:
```suggestion
      "EXPLAIN select * from t where (t.a, t.b) not in (select s.a, s.b from s); -- 3. anti semi join, two join key",
```

File: planner/core/plan_to_pb.go [hunk 6]
```diff
@@ -519,6 +536,7 @@ func (p *PhysicalHashJoin) ToPB(ctx sessionctx.Context, storeType kv.StoreType)
 		OtherConditions:         otherConditions,
 		OtherEqConditionsFromIn: otherEqConditions,
 		Children:                []*tipb.Executor{lChildren, rChildren},
+		IsNullAwareSemiJoin:     &isNullAwareSemiJoin,
```
Comment (AilinKid) [inline planner/core/plan_to_pb.go:539]: for golang，seems & can be removed

File: planner/core/rule_predicate_push_down.go [hunk 7]
```diff
@@ -322,29 +322,24 @@ func (p *LogicalJoin) updateEQCond() {
...
 					lNAKeys = append(lNAKeys, rExpr)
 					rNAKeys = append(rNAKeys, lExpr)
-					need2Remove = true
 				}
+				continue
 			}
-			if need2Remove {
-				p.OtherConditions = append(p.OtherConditions[:i], p.OtherConditions[i+1:]...)
-			}
+			otherCond = append(otherCond, p.OtherConditions[i])
```
Comment (AilinKid) [inline planner/core/rule_predicate_push_down.go:340]: this fixes the  reversed-key show，great～

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

**Review by windtalker — APPROVED** [review]
LGTM

**Review by AilinKid — APPROVED** [review]

File: planner/core/plan_to_pb.go (hunk 6 above)
Comment (gengliqi) [reply planner/core/plan_to_pb.go:539]: But `IsNullAwareSemiJoin` is a `*bool`.

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**